Split logging features of fastapi_log

SirPyTech opened the pull request:
As suggested in

> Thanks for working on this. I wanted to do it since a while :sweat_smile: 
> 
> Early feedback...
> 
> The way I was planning to handle this was to have a base module to collect any kind of "request/response log".
> This can be used by fastapi, base_rest, edi framework, endpoints and any custom implementation of request/response feature.
> 
> Would you consider splitting this part to a module like `request_log` or something like that?

_Originally posted by @simahawk in https://github.com/OCA/rest-framework/issues/501#issuecomment-2682063297_

Also add the module `fastapi_log_mail` to optionally send an email when an exception occurs.

Depends on:
 - [x] https://github.com/OCA/rest-framework/pull/515

lmignon committed:
[FIX] fastapi: Avoid zombie threads

Each time a fastapi app is created, a new event loop thread is created by the ASGIMiddleware. Unfortunately, every time the cache is cleared, a new app is created with a new even loop thread. This leads to an increase in the number of threads created to manage the asyncio event loop, even though many of them are no longer in use. To avoid this problem, the thread in charge of the event loop is now created only once per thread / process and the result is stored in the thread's local storage. If a new instance of an app needs to be created following a cache reset, this ensures that the same event loop is reused.

refs #484

lmignon committed:
[IMP] fastapi: add event loop lifecycle management

This commit adds event loop lifecycle management to the FastAPI dispatcher.

Before this commit, an event loop and the thread to run it were created
each time a FastAPI app was created. The drawback of this approach is that
when the app was destroyed (for example, when the cache of app was cleared),
the event loop and the thread were not properly stopped, which could lead
to memory leaks and zombie threads. This commit fixes this issue by creating
a pool of event loops and threads that are shared among all FastAPI apps.
On each call to a FastAPI app, a event loop is requested from the pool and
is returned to the pool when the app is destroyed. At request time of
an event loop, the pool try to reuse an existing event loop and if no event
loop is available, a new event loop is created.

The cache of the FastAPI app is also refactored to use it's own mechanism.
It's now based on a dictionary of queues by root path by database,
where each queue is a pool of FastAPI app. This allows a better management
of the invalidation of the cache. It's now possible to invalidate
the cache of FastAPI app by root path without affecting the cache of others
root paths.

lmignon committed:
[FIX] fastapi: Graceful shutdown of event loop

On server shutdown, ensure that created the event loops are closed properly.

lmignon committed:
[FIX] fastapi: Ensure thread safety of the FastAPI app cache

defaultdict in python is not thread safe. Since this data structure
is used to store the cache of FastAPI apps, we must ensure that the
access to this cache is thread safe. This is done by using a lock
to protect the access to the cache.

lmignon committed:
[IMP] fastapi: Improves app cache lifecycle

This commit improves the lifecycle of the fastapi app cache.
It first ensures that the cache is effectively invalidated when changes
are made to the app configuration even if theses changes occur into an
other server instance.
It also remove the use of a locking mechanism put in place to ensure a thread
safe access to a value into the cache to avoid potential concurrency issue when
a default value is set to the cache at access time. This lock could lead
to unnecessary contention and reduce the performance benefits of queue.Queue's
fine-grained internal synchronization for a questionable gain. The only
expected gain was to avoid the useless creation of a queue.Queue instance
that would never be used since at the time of puting the value into the cache
we are sure that a value is already present into the dictionary.

paradoxxxzero committed:
[FIX] extendable_fastapi: Inherit from last declared fastapi dispatcher

paradoxxxzero committed:
[ADD] fastapi_encrypted_errors

paradoxxxzero committed:
[FIX] fastapi: headers is not a dict, it can have multiple values for a key

PicchiSeba committed:
[IMP]fastapi: obtain endpoint via common method

lmignon committed:
[FIX] fastapi: remove useless code

OCA-git-bot committed:
Merge PR #486 into 16.0

Signed-off-by lmignon

OCA-git-bot committed:
[BOT] post-merge updates

lmignon committed:
[FIX] fastapi: Ensure no default eventloop thread is created

Before this change, an eventloop thread was created on each instanciation of the ASGIMiddleware class. We want to avoid this behavior to avoid the creation of zombie threads

OCA-git-bot committed:
Merge PR #504 into 16.0

Signed-off-by sebastienbeau

OCA-git-bot committed:
[BOT] post-merge updates

lmignon committed:
[IMP] base_rest_demo: Test to validate the retrying mechanism

Add a new test to ensure that if a retryable error occurs when processing a call to a base_rest endpoint, the retrying mechanism of odoo works as expected

lmignon committed:
[FIX] rest_log: Don't prevent the retrying mechanism

In case of a retryable error, the initial error must bubble up to the retrying mechanism. If this kind of error is wrapped into another one, the retrying mechanism no more works

OCA-git-bot committed:
Merge PR #508 into 16.0

Signed-off-by lmignon

OCA-git-bot committed:
[BOT] post-merge updates

OCA-git-bot committed:
Merge PR #500 into 16.0

Signed-off-by lmignon

OCA-git-bot committed:
[BOT] post-merge updates

SirPyTech committed:
[REF] fastapi_log: Extract common features to `api_log`

This way other APIs might use the new module `api_log` to store logs.

SirPyTech committed:
[ADD] fastapi_log_mail

PicchiSeba committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (43 of 43 strings)

Translation: rest-framework-16.0/rest-framework-16.0-fastapi
Translate-URL: https://translation.odoo-community.org/projects/rest-framework-16-0/rest-framework-16-0-fastapi/it/

PicchiSeba committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (3 of 3 strings)

Translation: rest-framework-16.0/rest-framework-16.0-fastapi_auth_jwt_demo
Translate-URL: https://translation.odoo-community.org/projects/rest-framework-16-0/rest-framework-16-0-fastapi_auth_jwt_demo/it/

SirPyTech committed:
[FIX] fastapi_log: Manage multi-slash endpoints

SirPyTech committed:
[IMP] api_log: Hide sensitive headers

PicchiSeba commented:
Regarding the [multi-slash PR](https://github.com/OCA/rest-framework/pull/515) maybe we should wait to include it in its entirety until we receive some feedback (and I implement some tests).

I could extract the `get_endpoint` method without the multi-slash part and put  it in its own PR and then apply my PR on it.
This way we have a common method to work on.

What do you think?

SirPyTech commented:
> Regarding the [multi-slash PR](https://github.com/OCA/rest-framework/pull/515) maybe we should wait to include it in its entirety until we receive some feedback (and I implement some tests).
> 
> I could extract the `get_endpoint` method without the multi-slash part and put it in its own PR and then apply my PR on it. This way we have a common method to work on.
> 
> What do you think?

Yes that's a good conservative solution.
So you'll create a simple refactoring PR that only extracts the occurrences of `self.search([("root_path", "=", root_path)])` in a `get_endpoint` method, leaving the rest unchanged; that should be merged easily.

Then https://github.com/OCA/rest-framework/pull/515 will build on the new PR to edit the `get_endpoint` implementation and `root_path` definition.

I will then include the new refactoring PR instead of https://github.com/OCA/rest-framework/pull/515.

PicchiSeba committed:
[IMP]fastapi: enable multi-slash routes

PicchiSeba commented:
@SirPyTech 

https://github.com/OCA/rest-framework/pull/524

I moved the common method to this PR

PicchiSeba commented:
@SirPyTech the commit 018f44d4a073620e85d615d84cce5aeb86e4e35d is missing.

SirPyTech commented:
> @SirPyTech the commit [018f44d](https://github.com/akretion/rest-framework/commit/018f44d4a073620e85d615d84cce5aeb86e4e35d) is missing.

That is not how PRs are included in OCA, see https://github.com/OCA/maintainer-tools/wiki/Use-temporary-reference(s)-to-another-pull-request(s).
To switch from https://github.com/OCA/rest-framework/pull/515 to https://github.com/OCA/rest-framework/pull/524 I pushed this change https://github.com/akretion/rest-framework/compare/6632ff858bcc9370e46d75853b4edfe7d1fc5314..641d4df4235e0c20f1efe482c926fecdd586d47d yesterday.

PicchiSeba commented:
> > @SirPyTech the commit [018f44d](https://github.com/akretion/rest-framework/commit/018f44d4a073620e85d615d84cce5aeb86e4e35d) is missing.
> 
> That is not how PRs are included in OCA, see https://github.com/OCA/maintainer-tools/wiki/Use-temporary-reference(s)-to-another-pull-request(s). To switch from [OCA#515](https://github.com/OCA/rest-framework/pull/515) to [OCA#524](https://github.com/OCA/rest-framework/pull/524) I pushed this change https://github.com/akretion/rest-framework/compare/6632ff858bcc9370e46d75853b4edfe7d1fc5314..641d4df4235e0c20f1efe482c926fecdd586d47d yesterday.

Oh thanks, I didn't know

PicchiSeba commented on api_log/models/api_log.py:
```diff
@@ -0,0 +1,213 @@
+        return time.time_ns() / 1e9
+
+    @api.model
+    def log_request(self, request):
+        log_request_values = {
+            "request_url": request.url,
+            "request_method": request.method,
+            "request_headers": self._headers_to_dict(request.headers),
+            "request_body": request.data,
+            "request_date": fields.Datetime.now(),
+            "request_time": self._current_time(),
+        }
```
**issue**: we should check if we can read `request.data` because if we don't and it is consumed then we get stuck into an infinite loop later on when calling `ASGIMiddleware.__call__()` (from https://github.com/OCA/rest-framework/blob/16.0/fastapi/middleware.py)

We could solve this like that

We change the method's signature to allow other parameters, then we move part of `fastapi_log.models.api_log`'s implementation into the original method

```suggestion
     def log_request(self, request, **kwargs):
        log_request_values = {
            "request_url": request.url,
            "request_method": request.method,
            "request_headers": self._headers_to_dict(request.headers),
            "request_date": fields.Datetime.now(),
            "request_time": self._current_time(),
        }
        stream = kwargs.get("fastapi_environ", {}).get("wsgi.input")
        if stream and stream.seekable():
            log_request_values["request_body"] = stream.read()
            stream.seek(0)
        else:
            log_request_values["request_body"] = request.data
```

An easier solution would have been to copy the `request` object but it doesn't have access to such a method

EDIT: nvm, `request` should have access to a method called `deepcopy`

SirPyTech replied:
Thanks for the proposal, I prefer to avoid adding `fastapi` code in the generic module `api_log` so I have added a couple of hooks, please check

PicchiSeba replied:
**praise**: that's actually a better way to do it, gj

PicchiSeba commented on fastapi_log/models/api_log.py:
```diff
@@ -0,0 +1,43 @@
+
+    @api.model
+    def log_request(self, request, fastapi_environ=None, fastapi_endpoint=None):
+        log = super().log_request(request)
```
2.
We leave the super method unchanged but we pass a copy of `request`

```suggestion
        log = super().log_request(request.copy())
```

PicchiSeba reviewed (changes requested):
Code review: issue when processing requests with consumable bodies

PicchiSeba commented on api_log/security/ir_model_access.xml:
```diff
@@ -5,10 +5,10 @@
-        <field name="name">Fastapi Log: Read access</field>
-        <field name="model_id" ref="model_fastapi_log" />
-        <field name="group_id" ref="group_fastapi_log" />
+    <record id="access_api_log" model="ir.model.access">
```
**question**: does this change require a migration?

SirPyTech replied:
Since `fastapi_log` isn't merged yet, I wouldn't bother doing migrations in general.

If we want to keep the user's edits on `fastapi_log`'s record then a migration is needed, otherwise Odoo will remove the old record and create a new one; since this is not a record usually edited by the user I wouldn't do a migration.

PicchiSeba commented on api_log/security/res_groups.xml:
```diff
@@ -6,8 +6,8 @@
 
-    <record id="group_fastapi_log" model="res.groups">
-        <field name="name">Fastapi Log Access</field>
+    <record id="group_api_log" model="res.groups">
```
**question**: (same as above)

SirPyTech replied:
See response for the question above.

PicchiSeba commented on api_log/tests/test_api_log.py:
```diff
@@ -0,0 +1,38 @@
+        request = requests.Request(
+            headers={
+                "Api-Key": secret_api_key,
+                "Cookie": secret_cookie,
```
**non-blocking**: let's add another header, we should check that it does not get deleted when logged.  For completeness

SirPyTech replied:
Right, done

PicchiSeba reviewed (commented):
Code review: good work! A couple of points

lmignon committed:
[FIX] datamodel: Not compatible with marshmallow >= 4.0.0

lmignon committed:
[IMP] fastapi: Ensures endpoint matching is based on path parts

When we lookup the fastapi.endpoint to use to process a request at a given path, we must ensure that the matching mechanism on the path and the endpoint root_path is bases on the path parts. e.g. /a/b is not prefix of /a/bc.
In the same time improves robustness of the matching mechanism and avoid useless SQL queries

OCA-git-bot committed:
Merge PR #527 into 16.0

Signed-off-by lmignon

OCA-git-bot committed:
[BOT] post-merge updates

OCA-git-bot committed:
Merge PR #520 into 16.0

Signed-off-by lmignon

OCA-git-bot committed:
[BOT] post-merge updates

PicchiSeba commented on fastapi_log/fastapi_dispatcher.py:
```diff
@@ -21,44 +21,36 @@ class FastApiDispatcher(_dispatchers.get("fastapi", BaseFastApiDispatcher)):
         fastapi_endpoint = (
             self.request.env["fastapi.endpoint"]
             .sudo()
-            .search([("root_path", "=", root_path)])
```
**issue**: the latest changes from @lmignon renamed this method to `_get_endpoint`, we should adapt its use

SirPyTech replied:
Done!
Actually, you did it :wink:
Thanks

PicchiSeba commented on fastapi_log/models/api_log.py:
```diff
@@ -0,0 +1,57 @@
+            log.fastapi_endpoint_id = (
+                request.env["fastapi.endpoint"]
+                .sudo()
+                .get_endpoint(environ["PATH_INFO"])
```
**issue**: same as above

SirPyTech replied:
Done!
Actually, you did it :wink:
Thanks

PicchiSeba reviewed (changes requested):
We changed a few things in the multi-slash PR. Let's adapt to them

PicchiSeba reviewed (approved):
Code and functional review: LGTM

anusriNPS reviewed (approved):
code and Functional Review: LGTM

jesusVMayor committed:
[FIX]base_rest: Fix rest controllers inheritance.

OCA-git-bot committed:
Merge PR #529 into 16.0

Signed-off-by lmignon

OCA-git-bot committed:
[BOT] post-merge updates

simahawk commented on api_log/models/api_log.py:
```diff
@@ -0,0 +1,224 @@
+        return httprequest.data
+
+    @api.model
+    def log_request(self, request):
```
This should accept custom values. Eg:
```suggestion
    def log_request(self, request, **values):
        log_request_values = {...}
        log_request_values.update(values)
```
This way you don't have to override the api.log model every time you implement a glue module.
Plus, it should use `sudo` to create.

SirPyTech replied:
> This way you don't have to override the api.log model every time you implement a glue module.

I prefer to override the method once than forcing the developer to remember to add the custom values each time this method is called; I've added them as suggested because they could be useful anyway.

>  Plus, it should use sudo to create.

I did let the caller choose if they wanted to use `sudo` or not, but now that you mention it I don't see why it should _not_ be used, so I added it here  as suggested and in other `write` calls.

simahawk commented on fastapi_log/models/api_log.py:
```diff
@@ -0,0 +1,57 @@
+        dispatcher = request.dispatcher
+        if dispatcher.routing_type == "fastapi":
+            environ = dispatcher._get_environ()
+            log.fastapi_endpoint_id = (
```
this should be passed by the call to `log_request` IMO

SirPyTech replied:
I understand your point of view, but I think that when `fastapi_log` is installed any call to `log_request` should simply do the right thing without the need to add special parameters to it: the `log_request` override recognizes that the `request` comes from `fastapi` and adds the required values automatically.

simahawk commented on fastapi_log/models/api_log.py:
```diff
@@ -0,0 +1,57 @@
+    )
+
+    @api.model
+    def _get_request_body(self, request):
```
this should be passed by the call to log_request IMO (see comment above)

SirPyTech replied:
That's a pain point because simply reading `httprequest.data` consumes it and breaks the flow of everyone else reading it later (whoever needs to manage the request for instance).

The implementation you proposed with the custom values wouldn't fix it because even if I read it first and then overwrite it with custom values, that still consumes the `httprequest.data` with the first read.

I know it's not very clean to extract only this value's computation but I couldn't find any cleaner solution, and it isn't necessary for other fields; let me know if you have a better idea!

simahawk commented on fastapi_log/models/api_log.py:
```diff
@@ -0,0 +1,57 @@
+class FastapiLog(models.Model):
+    _inherit = "api.log"
+
+    fastapi_endpoint_id = fields.Many2one(
```
I would prefer a generic field like I've done in `rest_log`. There we have `collection` and `collection_id`.

SirPyTech replied:
That's a good point, I created a collection model that is implemented from FastAPI endpoints, let me know what you think

simahawk commented on fastapi_log/models/fastapi_endpoint.py:
```diff
@@ -24,11 +24,13 @@ class FastapiEndpoint(models.Model):
-            [("endpoint_id", "in", self.ids)],
-            ["endpoint_id"],
-            ["endpoint_id"],
+        groups = self.env["api.log"].read_group(
```
IMO is not important to have a counter for log entries. You can have thousands... what's the point?

SirPyTech replied:
:+1:

simahawk commented on fastapi_log_mail/models/api_log.py:
```diff
@@ -0,0 +1,16 @@
+        res = super().log_exception(exception)
+        mail_template = self.fastapi_endpoint_id.fastapi_log_mail_template_id
+        if mail_template:
+            mail_template.sudo().send_mail(self.id)
```
Sending emails for every exception sounds too much.
I'd rather create an activity.

SirPyTech replied:
Good idea, thanks!
Since activities can send emails we are not losing the email feature anyway.

SirPyTech replied:
On a second thought the email is still needed because the activity assignation email is quite limited and does not allow a nice preview of the issue as the email does, hence I added both the email template and the activity type so the user can choose the preferred notification type :smile:

simahawk reviewed (commented):
Thanks guys! :)
You might want to add your name somewhere ;)

simahawk reviewed (changes requested):
some remarks

paradoxxxzero committed:
[ADD] auth_partner, fastapi_auth_partner

OCA-git-bot committed:
Merge PR #468 into 16.0

Signed-off-by sebastienbeau

Commit:
[UPD] Update fastapi_encrypted_errors.pot

OCA-git-bot committed:
[BOT] post-merge updates

OCA-git-bot committed:
Merge PR #456 into 16.0

Signed-off-by lmignon

Commit:
[UPD] Update auth_partner.pot

Commit:
[UPD] Update fastapi.pot

Commit:
[UPD] Update fastapi_auth_partner.pot

OCA-git-bot committed:
[BOT] post-merge updates

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: rest-framework-16.0/rest-framework-16.0-fastapi
Translate-URL: https://translation.odoo-community.org/projects/rest-framework-16-0/rest-framework-16-0-fastapi/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (43 of 43 strings)

Translation: rest-framework-16.0/rest-framework-16.0-fastapi
Translate-URL: https://translation.odoo-community.org/projects/rest-framework-16-0/rest-framework-16-0-fastapi/it/

mymage committed:
Added translation using Weblate (Italian)

mymage committed:
Added translation using Weblate (Italian)

lmignon committed:
[FIX] fastapi: Ensures inner exception is reset

OCA-git-bot committed:
Merge PR #533 into 16.0

Signed-off-by lmignon

OCA-git-bot committed:
[BOT] post-merge updates

mymage committed:
Added translation using Weblate (Italian)

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (19 of 19 strings)

Translation: rest-framework-16.0/rest-framework-16.0-fastapi_encrypted_errors
Translate-URL: https://translation.odoo-community.org/projects/rest-framework-16-0/rest-framework-16-0-fastapi_encrypted_errors/it/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (45 of 45 strings)

Translation: rest-framework-16.0/rest-framework-16.0-fastapi_auth_partner
Translate-URL: https://translation.odoo-community.org/projects/rest-framework-16-0/rest-framework-16-0-fastapi_auth_partner/it/

PicchiSeba committed:
[IMP]fastapi: add test for multi-slash scenario

grindtildeath committed:
[FIX] fastapi: Flush before exiting Fastapi environment

As the fastapi dispatcher is altering the public request's
environment before calling the endpoint, it will reuse
the same DB cursor from this environment.

Therefore, if fields are being marked to be recomputed during
the execution of the fastapi endpoint, these will not be
recomputed until the cursor is being commited, what happens
when we are not anymore authenticated through as fastapi user.
This means the recomputation of computed fields will be executed
with the public user, what could potentially lead to inconsistencies
or result in an AccessError.

By calling flush before exiting the contextmanager yielding
the altered Odoo environment, we ensure all the pending
computations are executed while being authenticated with
the fastapi user which must have more access rights than the
public user.

Tisho99 committed:
[FIX] fastapi: Running User Rule Domain

OCA-git-bot committed:
Merge PR #542 into 16.0

Signed-off-by lmignon

OCA-git-bot committed:
[BOT] post-merge updates

OCA-git-bot committed:
Merge PR #545 into 16.0

Signed-off-by lmignon

OCA-git-bot committed:
[BOT] post-merge updates

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (90 of 90 strings)

Translation: rest-framework-16.0/rest-framework-16.0-auth_partner
Translate-URL: https://translation.odoo-community.org/projects/rest-framework-16-0/rest-framework-16-0-auth_partner/it/

OCA-git-bot committed:
Merge PR #515 into 16.0

Signed-off-by lmignon

OCA-git-bot committed:
[BOT] post-merge updates

SirPyTech committed:
[IMP] api_log: Add collection of logs

SirPyTech committed:
[ADD] api_log_mail: Notify user about logged exceptions

SirPyTech committed:
[IMP] fastapi_log: Adapt to log collection

SirPyTech committed:
[IMP] fastapi_log_mail: Adapt to log collection

SirPyTech commented:
There are several conflicts because https://github.com/akretion/rest-framework/tree/16.0-add-fastapi_log is not updated with https://github.com/OCA/rest-framework/tree/16.0, maybe it's time to move this PR to OCA?

simahawk commented:
> There are several conflicts because https://github.com/akretion/rest-framework/tree/16.0-add-fastapi_log is not updated with https://github.com/OCA/rest-framework/tree/16.0, maybe it's time to move this PR to OCA?

Yes please :)

simahawk commented on api_log/models/api_log_collection.py:
```diff
@@ -0,0 +1,46 @@
+
+
+class APILogCollection(models.AbstractModel):
+    _name = "api.log_collection"
```
```suggestion
    _name = "api.log_collection.mixin"
```
If I get right your intent.

simahawk commented:
Also, regarding autorship of `api_log`... the design of this thing comes from rest_log which I made w/ Guewen: could you please add ourselves as contributors at least?

paradoxxxzero committed:
[ADD] fastapi_log

SirPyTech commented:
Moved to https://github.com/OCA/rest-framework/pull/554.
